acrn: cannot start predefined ACRN domain

tttech-industrial-buchsbaum opened the pull request:
Starting a predefined ACRN VM fails:

```
root@nuc7i7dnh:~# virsh list --all
 Id   Name            State
--------------------------------
 -    DebianBuster0   shut off

root@nuc7i7dnh:~# virsh start DebianBuster0
error: Failed to start domain DebianBuster0
error: internal error: vm(00000000-0000-0000-0000-000000000000) not found

```
The libvirtd log shows

```
libvirtd: 1623: error : acrnDomainCreateWithFlags:1619 : internal error: vm(00000000-0000-0000-0000-000000000000) not found
```
Fix this by populating private data hvUUID for all predefined ACRN domains
in acrnStateInitialize().

Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2557,6 +2557,19 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
     return ret;
 }
 
+static
```
Please put static and void in one line

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2557,6 +2557,19 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
+
+        /* populate hvUUID */
+        priv = domObj->privateData;
+        uuid_copy(priv->hvUUID, domObj->def->uuid);
```
hvUUID is different from def->uuid and needs to be acquired via acrnAllocateVm(). Please refer to acrnDomainDefineXMLFlags() for its usage.

tttech-industrial-buchsbaum replied:
The updated patch now is based on acrnAllocateVm().

tttech-industrial-buchsbaum committed:
acrn: cannot start predefined ACRN domain

Starting a predefined ACRN VM fails:

root@nuc7i7dnh:~# virsh list --all
Id   Name            State
--------------------------------
-    DebianBuster0   shut off

root@nuc7i7dnh:~# virsh start DebianBuster0
error: Failed to start domain DebianBuster0
error: internal error: vm(00000000-0000-0000-0000-000000000000) not found

The libvirtd log shows

libvirtd: 1623: error : acrnDomainCreateWithFlags:1619 : internal error: vm(00000000-0000-0000-0000-000000000000) not found

Fix this by allocating a VM for each predefined ACRN domain in
acrnStateInitialize().

Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2605,6 +2650,9 @@ acrnStateInitialize(bool privileged,
         goto cleanup;
 
+    virDomainObjListForEach(acrn_driver->domains, acrnPersistentDomainInit,
+        acrn_driver);
```
If the return value is < 0, we should go to cleanup.

peterfang replied:
A side comment is we're doing something potentially dangerous here, which stemmed from the implementation of acrnAllocateVm(). We hold the domain's read lock recursively during this call, but we're actually changing its content (priv->hvUUID). However, pthread_rwlock_wrlock() does not define the behavior of recursive write locks or lock upgrades. This only happens during acrnStateInitialize() right now, so we're not in danger of a race condition. But it's something we should be aware of.

tttech-industrial-buchsbaum replied:
Using virDomainObjListForEach without checking the return value seems a common practice throughout libvirt and I think this is why I just missed it. But you are perfectly right, I'll add that check.

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2556,6 +2556,50 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
+    uuid_copy(priv->hvUUID, hvUUID);
+
+    event = virDomainEventLifecycleNewFromObj(dom,
+                                              VIR_DOMAIN_EVENT_DEFINED,
```
Let's make this a bit more robust here. We can goto cleanup if !event.

tttech-industrial-buchsbaum replied:
Again, checking the event pointer and bailing out with an error seem correct. Will be in the next patch ...

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2556,6 +2556,50 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
+    if (!vmList)
+        goto cleanup;
+
+    acrnDriverLock(privconn);
```
Please remove this. Since the domain's rw lock is held when this function is called, calling acrnDriverLock() creates a potential deadlock.

tttech-industrial-buchsbaum replied:
Removing the lock/unlock in `acrnPersistentDomainInit` should not result in any problems, since here we are in the initialization phase of the driver. The problem obviously is holding a domain lock and then acquiring the driver lock. What about removing lock/unlock in `acrnPersistentDomainInit` and acquire the driver lock outside the iteration:
```
acrnDriverLock(privconn);
virDomainObjListForEach(acrn_driver->domains, ...)
acrnDriverUnlock(privconn);
```

It took my me some additional steps through the code to understand your point about the recursive locking (using `virDomainObjListForEach` in `acrnPersistentDomainInit` and recursively in `acrnAllocateVm`). Are there any chances for a clean way out of this? Probably adding a 'lockless' variant of `acrnAllocateVm` using `virHashForEach` directly protected just by the driver lock?  Or just doing this in `acrnAllocateVm` as it is? Since right now the driver lock is held by every caller of `acrnAllocateVm` using `virHashForEach` is possible, it means just adding the requirement to hold the driver lock when calling `acrnAllocateVm` for any future user. What do you think about it?

tttech-industrial-buchsbaum replied:
Just minutes later, the above seems a bad idea since this requires the driver lock to be held everywhere the domain list gets manipulated!

tttech-industrial-buchsbaum replied:
Just found this in the manpage of pthread_rwlock_rdlock:

> A thread may hold multiple concurrent read locks on rwlock (that is, successfully call the pthread_rwlock_rdlock() function n times). If so, the application shall ensure that the thread performs matching unlocks (that is, it calls the pthread_rwlock_unlock() function n times).

peterfang replied:
We could try to have a lockless flavor of `acrnAllocateVm`. Right now, it would only be used in `acrnStateInitialize` so we don't need to grab the driver lock at all (`acrn_driver` is not yet observable to anyone other than this function). In this case, I'd actually prefer your first version using `notify` in `virDomainObjListLoadAllConfigs`. The domain's write lock will be held during `notify`, so it can't be acquired (read or write) again in `acrnAllocateVm`.

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2556,6 +2556,50 @@ acrnInitPlatform(acrnPlatformInfoPtr pi, virNodeInfoPtr nodeInfo,
+    ret = 0;
+
+cleanup_unlock:
+    acrnDriverUnlock(privconn);
```
Please remove this as well.

tttech-industrial-buchsbaum commented:
PATCHv3 changes:
* Removed the driver lock
* Calling `acrnGetPlatform` outside the iteration and thus only once

I am still using the `virDomainObjListForEach` iterator (with the recursive, but harmless read locking of the domain list lock, locked by `pthread_rwlock_rdlock`). A notify function of `virDomainObjListLoadAllConfigs` gets blocked in `arcnVmAllocate` since `virDomainObjListLoadAllConfigs` holds the domain list lock by `pthread_rwlock_wrlock`!

tttech-industrial-buchsbaum committed:
acrn_driver: avoid duplicate read of ACRN platform data during init

Since the ACRN vm list has to be obtained in acrnStateInitialize() for
iterating through all domains, reuse this list for acrnInitPlatform()
to avoid unnecessary duplicate read of ACRN platform data during
initialization.

Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>

tttech-industrial-buchsbaum committed:
acrn_driver: introduce acrnIsRtvm

Since rtvm property lives in ACRN specific namespace data which is stored
as opaque namespaceData pointer in the domain definition, it always can
be derived from the actual domain definition. Provide an accessor to
the rtvm property via domain definition and use it wherever possible.
Furthermore use this acrnAllocateVm() which already uses the domain
definition and thus can access the rtvm property without the need to pass
it as parameter.

Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>

peterfang commented on src/acrn/acrn_driver.c:
```diff
@@ -2792,8 +2783,15 @@ acrnStateInitialize(bool privileged,
+    if (!list)
+        goto cleanup;
+
+    if (acrnGetPlatform(&acrn_driver->pi, list) < 0)
```
-ENODEV needs to be handled here

peterfang reviewed (commented):
This is a great change!

peterfang reviewed (commented):
This looks good. Thanks!

tttech-industrial-buchsbaum commented:
Sorry for taking so much time for the follow-up patch. But today I just push V4 of my patches:
Patchv4 changes:

* correct the return value handling for acrnGetPlatform

peterfang commented:
Looks great! Thanks for the hard work @tttech-industrial-buchsbaum

peterfang commented:
I will port this to the `dev-acrn-v6.1.0` branch.
